plugins/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobContainerRetriesTests.java

```diff
@@ -392,6 +395,66 @@ public void testWriteLargeBlob() throws IOException {
         assertThat(allow410Gone.get(), is(false));
     }
 
+    public void testDeleteBatchesAreSentIncrementally() throws Exception {
+        // See com.google.cloud.storage.spi.v1.HttpStorageRpc.DefaultRpcBatch.MAX_BATCH_SIZE
+        final int sdkMaxBatchSize = 100;
+        final AtomicInteger receivedBatchRequests = new AtomicInteger();
+
+        final int totalDeletes = randomIntBetween(MAX_DELETES_PER_BATCH - 1, MAX_DELETES_PER_BATCH * 2);
+        final AtomicInteger pendingDeletes = new AtomicInteger();
+        final Iterator<String> blobNamesIterator = new Iterator<>() {
+            int totalDeletesSent = 0;
+
+            @Override
+            public boolean hasNext() {
+                return totalDeletesSent < totalDeletes;
+            }
+
+            @Override
+            public String next() {
+                if (pendingDeletes.get() == MAX_DELETES_PER_BATCH) {
+                    // Check that once MAX_DELETES_PER_BATCH deletes are enqueued the pending batch requests are sent
+                    assertThat(receivedBatchRequests.get(), is(greaterThan(0)));
+                    assertThat(receivedBatchRequests.get(), is(lessThanOrEqualTo(MAX_DELETES_PER_BATCH / sdkMaxBatchSize)));
+                    receivedBatchRequests.set(0);
+                    pendingDeletes.set(0);
+                }
+
+                pendingDeletes.incrementAndGet();
+                return Integer.toString(totalDeletesSent++);
+            }
+        };
+        final BlobContainer blobContainer = createBlobContainer(1, null, null, null);
+        httpServer.createContext("/batch/storage/v1", safeHandler(exchange -> {
+            assert pendingDeletes.get() <= MAX_DELETES_PER_BATCH;
+
+            receivedBatchRequests.incrementAndGet();
+            final StringBuilder batch = new StringBuilder();
+            for (String line : Streams.readAllLines(exchange.getRequestBody())) {
+                if (line.length() == 0 || line.startsWith("--") || line.toLowerCase(Locale.ROOT).startsWith("content")) {
+                    batch.append(line).append("\r\n");
+                } else if (line.startsWith("DELETE")) {
+                    batch.append("HTTP/1.1 204 NO_CONTENT").append("\r\n");
+                    batch.append("\r\n");
+                }
+            }
+            byte[] response = batch.toString().getBytes(UTF_8);
+            exchange.getResponseHeaders().add("Content-Type", exchange.getRequestHeaders().getFirst("Content-Type"));
+            exchange.sendResponseHeaders(RestStatus.OK.getStatus(), response.length);
+            exchange.getResponseBody().write(response);
+        }));
+
+        blobContainer.deleteBlobsIgnoringIfNotExists(blobNamesIterator);
+
+        // Ensure that the remaining deletes are sent in the last batch
+        if (pendingDeletes.get() > 0) {
```

Comment:
Can we also assert that no extra request is sent/received when `pendingDeletes.get() == 0`?